src/backend/cpu/kernel/meanshift.hpp

```diff
@@ -11,125 +11,133 @@
                     dim_t i_out_off = i*ostrides[0];
 
-                    // clear means and centers for this pixel
-                    for(dim_t ch=0; ch<channels; ++ch) {
-                        means[ch] = 0.0f;
-                        // the expression ch*istrides[2] will only effect when ch>1
-                        // i.e for color images where batch is along fourth dimension
+                    std::vector<T> centers(channels, 0);
+
+                    for (unsigned ch=0; ch<channels; ++ch)
                         centers[ch] = inData[j_in_off + i_in_off + ch*istrides[2]];
-                    }
 
-                    // scope of meanshift iterationd begin
-                    for(unsigned it=0; it<iter; ++it) {
+                    int cj = j;
+                    int ci = i;
 
-                        int count   = 0;
-                        int shift_x = 0;
+                    // scope of meanshift iterations begin
+                    for (unsigned it=0; it<iter; ++it) {
+
+                        int ocj   = cj;
+                        int oci   = ci;
+                        unsigned count = 0;
                         int shift_y = 0;
+                        int shift_x = 0;
+
+                        std::vector<AccType> means(channels, 0);
 
-                        for(dim_t wj=-radius; wj<=radius; ++wj) {
+                        // Windowing operation
+                        for (dim_t wj=-radius; wj<=radius; ++wj) {
 
                             int hit_count = 0;
+                            dim_t tj = cj + wj;
+
+                            if (tj<0 || tj>dims[1]-1) break;
+
+                            dim_t tjstride = tj*istrides[1];
+
+                            for (dim_t wi=-radius; wi<=radius; ++wi) {
 
-                            for(dim_t wi=-radius; wi<=radius; ++wi) {
+                                dim_t ti = ci + wi;
 
-                                dim_t tj = j + wj;
-                                dim_t ti = i + wi;
+                                if (ti<0 || ti>dims[0]-1) break;
 
-                                // clamps offsets
-                                tj = clamp(tj, 0ll, dims[1]-1);
-                                ti = clamp(ti, 0ll, dims[0]-1);
+                                dim_t tistride = ti*istrides[0];
 
-                                // proceed
-                                float norm = 0.0f;
-                                for(dim_t ch=0; ch<channels; ++ch) {
-                                    tmpclrs[ch] = inData[ tj*istrides[1] + ti*istrides[0] + ch*istrides[2]];
-                                    norm += (centers[ch]-tmpclrs[ch]) * (centers[ch]-tmpclrs[ch]);
+                                std::vector<T> tmpclrs(channels, 0);
+
+                                AccType norm = 0;
+                                for (unsigned ch=0; ch<channels; ++ch) {
+                                    tmpclrs[ch] = inData[ tistride + tjstride + ch*istrides[2] ];
```

Comment:
The value of `channels` can be only 3 or 1, and the input validation checks for this are in place under `src/api/c/meanshift.cpp` file already. Also, each color channel contributes to the `norm` variable which is in term used to calculate the mean position and color. I am not sure this is causing the problem you think it is.

Reply:
I don't think this is an input issue. I understand that you can only have one or three channels. In the one channel scenario, this will not be a problem because the access will be linear. This will be a pretty big problem when we are dealing with a 3 channel image because you are accessing data down the third dimension. This type of accesses are not problems in CUDA or OpenCL because each access is linear wrt the threads.

A better approach would be to read the data into a std::array and perform the norm on it before doing a write. This ensures that the reads will be contiguous along the first dimension. Test with a few different sizes to get the optimal value.